File: go/2nd_ex/main.go
package main

import (
	"fmt"
)

func initVal() int {
	var val int
	_, err := fmt.Scanf("%d", &val)
	if err != nil || val <= 0 {
		panic("Invalid input")
	}
	return val
}

func cmp(arr []int, v int) {
	maxVal := 0
	for _, val := range arr {
		if val > maxVal {
			maxVal = val
		}
	}
	if v < maxVal {
		panic("The volume of the watering can is not enough to water the flower")
	}
}

func countingSteps(arr []int, v int) int {
	count := 0
	tmp := v

	for i, val := range arr {
		if tmp < val {
			tmp = v
			count += i * 2
		}
		count++
		tmp -= val
	}

	return count
}

func main() {
	n := initVal()

	arr := make([]int, n)
	for i := 0; i < n; i++ {
		arr[i] = initVal()
	}

	v := initVal()

	cmp(arr, v)

	fmt.Println(countingSteps(arr, v))
}

File: swift/2nd_ex/main.swift
import Foundation

func err(_ exc: String) -> Never {
    fatalError(exc)
}

func initVal() -> Int {
    guard let line = readLine(), let val = Int(line), val > 0 else {
        err("Invalid input")
    }
    return val
}

func cmp(_ arr: [Int], _ v: Int) {
    guard v >= arr.max() ?? 0 else {
        err("The volume of the watering can is not enough to water the flower")
    }
}

func countingSteps(_ arr: [Int], _ v: Int) -> Int {
    var count = 0
    var tmp = v

    for (i, val) in arr.enumerated() {
        if tmp < val {
            tmp = v
            count += i * 2
        }
        count += 1
        tmp -= val
    }

    return count
}

let n = initVal()
var arr: [Int] = []

for _ in 0..<n {
    arr.append(initVal())
}

let v = initVal()

cmp(arr, v)

print(countingSteps(arr, v))
